arangod/Cluster/Maintenance.cpp

```diff
@@ -618,13 +661,19 @@ arangodb::Result arangodb::maintenance::executePlan(VPackSlice const& plan,
           << "adding action " << action.get() << " to feature ";
       VPackObjectBuilder b(&report);
       action->toVelocyPack(report);
-      feature.addAction(std::move(action), false);
+      Result res = feature.addAction(std::move(action), false);
+      if (res.fail()) {
+        feature.unlockShardByAction(action);
+      }
```

Comment:
Is this race-free?
For example, if no shard lock was place in the map of shard locks by addAction(), but addAction() returns a failure, then we still remove the shard lock from the locks map here. Is it possible that we remove another action's shard lock then?

Reply:
I am also a bit concerned that a shard lock may never be cleaned up if an exception happens after a shard lock was acquired. Then the lock would stay around forever, and no action on the shard would be carried out

Reply:
The problem is that the lock must be acquired by phaseOne in one thread and must (normally) be released by the termination of the action in another thread. So scope guards cannot be used. The `addAction` is only called when we just got the lock for this shard, and this is only done if the lock was not held by anybody. However, if `addAction` somehow fails, we must make sure that the lock is released. This is what this code is good for here.